
<!DOCTYPE html>
<html>
  <head>
    <title>Communities API API</title>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <script type="text/javascript" src="bootstrap.min.js"></script>
    <link rel="stylesheet" type="text/css" href="bootstrap.min.css" />
  </head>

  <body style="background-color: white; margin: 15px;">
    <div class="navbar navbar-inverse navbar-fixed-top">
      <div class="navbar-inner">
	<div class="container">
	  <a class="brand" href="./index.html">Communities API API documentation</a>
	</div>
      </div>
    </div>

    <div class="row" style="margin-top: 50px;">
      <div name="index" class="span3" style="position: fixed; width: 130px;">
	<ul class="nav nav-list">
	  <li class="nav-header">Resources</li>
<li><a href="#overview">overview</a></li><li><a href="#annotation">annotation</a></li><li><a href="#compute">compute</a></li><li><a href="#download">download</a></li><li><a href="#inbox">inbox</a></li><li><a href="#library">library</a></li><li><a href="#m5nr">m5nr</a></li><li><a href="#matrix">matrix</a></li><li><a href="#metadata">metadata</a></li><li><a href="#metagenome">metagenome</a></li><li><a href="#project">project</a></li><li><a href="#sample">sample</a></li></ul>	
    </div>
    <div class="span12" style="margin-left: 270px;">
<h1><a name='overview' style='padding-top: 50px;'>Communities API Overview</a></h1><p>RESTful Metagenomics RAST object and resource API
For usage note that required parameters need to be passed as path parameters, optional parameters need to be query parameters. If an optional parameter has a list of option values, the first displayed will be used as default.</p><hr><h1><a name='annotation' style='padding-top: 50px;'>annotation</a></h1><h2>Description</h2><p>All annotations of a metagenome for a specific annotation type and source</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>sequence - request</h2><h3>Description</h3><p>tab deliminted annotated sequence stream</p><p>This is a stream GET request.</p><h3>Parameters</h3><ul><li><b>evalue</b> (int)</p><p>negative exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</p><p>text string to filter annotations by: only return those that contain text</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_level</b> (string)</p><p>hierarchal level to filter annotations by, for organism or ontology only</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</p><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</p><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RefSeq</b> - protein database, type organism, function, feature</li><li><b>GenBank</b> - protein database, type organism, function, feature</li><li><b>IMG</b> - protein database, type organism, function, feature</li><li><b>SEED</b> - protein database, type organism, function, feature</li><li><b>TrEMBL</b> - protein database, type organism, function, feature</li><li><b>SwissProt</b> - protein database, type organism, function, feature</li><li><b>PATRIC</b> - protein database, type organism, function, feature</li><li><b>KEGG</b> - protein database, type organism, function, feature</li><li><b>RDP</b> - RNA database, type organism, function, feature</li><li><b>Greengenes</b> - RNA database, type organism, function, feature</li><li><b>LSU</b> - RNA database, type organism, function, feature</li><li><b>SSU</b> - RNA database, type organism, function, feature</li><li><b>Subsystems</b> - ontology database, type ontology only</li><li><b>NOG</b> - ontology database, type ontology only</li><li><b>COG</b> - ontology database, type ontology only</li><li><b>KO</b> - ontology database, type ontology only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>organism</b> - return organism data</li><li><b>function</b> - return function data</li><li><b>ontology</b> - return ontology data</li><li><b>feature</b> - return feature data</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/annotation/sequence/mgm4447943.3?evalue=10&type=organism&source=SwissProt<li>all annotated read sequences from mgm4447943.3 with hits in SwissProt organisms at evaule < e-10</ul><h3>Return Attributes</h3><ul><li><b>streaming text</b> (object)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>col_01 (string)</b> - sequence id</li><li><b>col_02 (string)</b> - m5nr id (md5sum)</li><li><b>col_03 (string)</b> - semicolon seperated list of annotations</li><li><b>col_04 (string)</b> - dna sequence</li></ul><br></p></ul><h2>similarity - request</h2><h3>Description</h3><p>tab deliminted blast m8 with annotation</p><p>This is a stream GET request.</p><h3>Parameters</h3><ul><li><b>evalue</b> (int)</p><p>negative exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</p><p>text string to filter annotations by: only return those that contain text</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_level</b> (string)</p><p>hierarchal level to filter annotations by, for organism or ontology only</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</p><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</p><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RefSeq</b> - protein database, type organism, function, feature</li><li><b>GenBank</b> - protein database, type organism, function, feature</li><li><b>IMG</b> - protein database, type organism, function, feature</li><li><b>SEED</b> - protein database, type organism, function, feature</li><li><b>TrEMBL</b> - protein database, type organism, function, feature</li><li><b>SwissProt</b> - protein database, type organism, function, feature</li><li><b>PATRIC</b> - protein database, type organism, function, feature</li><li><b>KEGG</b> - protein database, type organism, function, feature</li><li><b>RDP</b> - RNA database, type organism, function, feature</li><li><b>Greengenes</b> - RNA database, type organism, function, feature</li><li><b>LSU</b> - RNA database, type organism, function, feature</li><li><b>SSU</b> - RNA database, type organism, function, feature</li><li><b>Subsystems</b> - ontology database, type ontology only</li><li><b>NOG</b> - ontology database, type ontology only</li><li><b>COG</b> - ontology database, type ontology only</li><li><b>KO</b> - ontology database, type ontology only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>organism</b> - return organism data</li><li><b>function</b> - return function data</li><li><b>ontology</b> - return ontology data</li><li><b>feature</b> - return feature data</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/annotation/similarity/mgm4447943.3?identity=80&type=function&source=KO<li>all annotated read blat stats from mgm4447943.3 with hits in KO functions at % identity > 80</ul><h3>Return Attributes</h3><ul><li><b>streaming text</b> (object)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>col_01 (string)</b> - query sequence id</li><li><b>col_02 (string)</b> - hit m5nr id (md5sum)</li><li><b>col_03 (float)</b> - percentage identity</li><li><b>col_04 (int)</b> - alignment length,</li><li><b>col_05 (int)</b> - number of mismatches</li><li><b>col_06 (int)</b> - number of gap openings</li><li><b>col_07 (int)</b> - query start</li><li><b>col_08 (int)</b> - query end</li><li><b>col_09 (int)</b> - hit start</li><li><b>col_10 (int)</b> - hit end</li><li><b>col_11 (float)</b> - e-value</li><li><b>col_12 (float)</b> - bit score</li><li><b>col_13 (string)</b> - semicolon seperated list of annotations</li></ul><br></p></ul><hr><h1><a name='compute' style='padding-top: 50px;'>compute</a></h1><h2>Description</h2><p>Calculate a PCoA for given input data.</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>normalize - request</h2><h3>Description</h3><p>Calculate normalized values for given input data.</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>columns</b> (list of string)</li><p>column id</p><p>This parameter must be passed in the message body.</p><li><b>data</b> (list of list)</li><p>raw value</p><p>This parameter must be passed in the message body.</p><li><b>rows</b> (list of string)</li><p>row id</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"columns":["mgm4441619.3","mgm4441656.4"],"rows":["Eukaryota","Bacteria","Archaea"],"data":[[135,410],[4397,6529],[1422,2156]]}' "http://kbase.us/services/communities/1/compute/normalize"<li>retrieve normalized values for inputed abundaces</ul><h3>Return Attributes</h3><ul><li><b>columns</b> (list of strings)</li><p>column id</p><li><b>data</b> (list of lists of floats)</li><p>normalized value</p><li><b>rows</b> (list of strings)</li><p>row id</p></p></ul><h2>heatmap - request</h2><h3>Description</h3><p>Calculate a dendogram for given input data.</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>cluster</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>ward</b> - ward cluster method</li><li><b>single</b> - single cluster method</li><li><b>complete</b> - complete cluster method</li><li><b>mcquitty</b> - mcquitty cluster method</li><li><b>median</b> - median cluster method</li><li><b>centroid</b> - centroid cluster method</li></ul><br><p>This parameter must be passed in the message body.</p><li><b>columns</b> (list of string)</li><p>column id</p><p>This parameter must be passed in the message body.</p><li><b>data</b> (list of list)</li><p>raw or normalized value</p><p>This parameter must be passed in the message body.</p><li><b>distance</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>bray-curtis</b> - bray-curtis distance method</li><li><b>euclidean</b> - euclidean distance method</li><li><b>maximum</b> - maximum distance method</li><li><b>manhattan</b> - manhattan distance method</li><li><b>canberra</b> - canberra distance method</li><li><b>minkowski</b> - minkowski distance method</li><li><b>difference</b> - difference distance method</li></ul><br><p>This parameter must be passed in the message body.</p><li><b>raw</b> (boolean)</p><p>option to use raw data (not normalize)</p><p>This parameter must be passed in the message body.</p><li><b>rows</b> (list of string)</li><p>row id</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"raw":0,"cluster":"mcquitty""columns":["mgm4441619.3","mgm4441656.4"],"rows":["Eukaryota","Bacteria","Archaea"],"data":[[135,410],[4397,6529],[1422,2156]]}' "http://kbase.us/services/communities/1/compute/heatmap"<li>retrieve dendogram of normalized inputed abundances using 'mcquitty' cluster method</ul><h3>Return Attributes</h3><ul><li><b>coldend</b> (object)</li><p>dendogram object for columns</p><li><b>colindex</b> (list of floats)</li><p>column id index</p><li><b>columns</b> (list of strings)</li><p>column id</p><li><b>data</b> (list of lists of floats)</li><p>normalized value</p><li><b>rowdend</b> (object)</li><p>dendogram object for rows</p><li><b>rowindex</b> (list of floats)</li><p>row id index</p><li><b>rows</b> (list of strings)</li><p>row id</p></p></ul><h2>pcoa - request</h2><h3>Description</h3><p>Calculate a PCoA for given input data.</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>columns</b> (list of string)</li><p>column id</p><p>This parameter must be passed in the message body.</p><li><b>data</b> (list of list)</li><p>raw or normalized value</p><p>This parameter must be passed in the message body.</p><li><b>distance</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>bray-curtis</b> - bray-curtis distance method</li><li><b>euclidean</b> - euclidean distance method</li><li><b>maximum</b> - maximum distance method</li><li><b>manhattan</b> - manhattan distance method</li><li><b>canberra</b> - canberra distance method</li><li><b>minkowski</b> - minkowski distance method</li><li><b>difference</b> - difference distance method</li></ul><br><p>This parameter must be passed in the message body.</p><li><b>raw</b> (boolean)</p><p>option to use raw data (not normalize)</p><p>This parameter must be passed in the message body.</p><li><b>rows</b> (list of string)</li><p>row id</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"raw":1,"distance":"euclidean""columns":["mgm4441619.3","mgm4441656.4"],"rows":["Eukaryota","Bacteria","Archaea"],"data":[[135,410],[4397,6529],[1422,2156]]}' "http://kbase.us/services/communities/1/compute/pcoa"<li>retrieve PCO of raw inputed abundances using 'euclidean' distance method</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - column id</li><li><b>pco (list of floats)</b> - principal component value</li></ul><br><li><b>pco</b> (list of floats)</li><p>average principal component value</p></p></ul><hr><h1><a name='download' style='padding-top: 50px;'>download</a></h1><h2>Description</h2><p>An analysis file from the processing of a metagenome from a specific stage in its analysis</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>instance - request</h2><h3>Description</h3><p>Returns a single sequence file.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>file</b> (string)</p><p>file name or identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/download/mgm4447943.3?file=350.1<li>download fasta file of genecalled protein sequences (from stage 350)</ul><h3>Return Attributes</h3><ul><li><b>data</b> (file)</li><p>requested analysis file</p></p></ul><h2>setlist - request</h2><h3>Description</h3><p>Returns a list of sets of sequence files for the given id.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>stage</b> (string)</p><p>stage name or identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/download/mgm4447943.3?stage=650<li>view all available files from stage 650</ul><h3>Return Attributes</h3><ul><li><b>file_id</b> (string)</li><p>unique identifier of file in stage</p><li><b>file_name</b> (string)</li><p>name of the analysis file</p><li><b>id</b> (string)</li><p>unique metagenome identifier</p><li><b>stage_id</b> (string)</li><p>three digit numerical identifier of the stage</p><li><b>stage_name</b> (string)</li><p>name of the stage in processing of this file</p><li><b>stage_type</b> (string)</li><p>type of the analysis file within a stage, i.e. passed or removed for quality control steps</p><li><b>url</b> (string)</li><p>url for retrieving this analysis file</p></p></ul><hr><h1><a name='inbox' style='padding-top: 50px;'>inbox</a></h1><h2>Description</h2><p>inbox receives user inbox data upload, requires authentication, see http://blog.metagenomics.anl.gov/mg-rast-v3-2-faq/#api_submission for details</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>view - request</h2><h3>Description</h3><p>lists the contents of the user inbox</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>auth</b> (string)</p><p>unique string of text generated by MG-RAST for your account</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>curl -X GET -H "auth: auth_key" "http://kbase.us/services/communities/1/inbox"<li>lists the contents of the user inbox, auth is required</ul><h3>Return Attributes</h3><ul><li><b>files</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>filename (string)</b> - path of file from within user inbox</li><li><b>filesize (string)</b> - disk size of file in bytes</li><li><b>timestamp (string)</b> - timestamp of file</li></ul><br><li><b>id</b> (string)</li><p>user login</p><li><b>timestamp</b> (string)</li><p>timestamp for return of this query</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p></p></ul><h2>upload - request</h2><h3>Description</h3><p>receives user inbox data upload</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>auth</b> (string)</p><p>unique string of text generated by MG-RAST for your account</p><p>This parameter must be passed in the message body.</p><li><b>upload</b> (file)</p><p>file to upload to inbox</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -H "auth: auth_key" -F "upload=@sequences.fastq" "http://kbase.us/services/communities/1/inbox"<li>upload file 'sequences.fastq' to user inbox, auth is required</ul><h3>Return Attributes</h3><ul><li><b>id</b> (string)</li><p>user login</p><li><b>status</b> (string)</li><p>status message</p><li><b>timestamp</b> (string)</li><p>timestamp for return of this query</p></p></ul><hr><h1><a name='library' style='padding-top: 50px;'>library</a></h1><h2>Description</h2><p>A library of metagenomic samples from some environment</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/library?limit=20&order=name<li>retrieve the first 20 libraries ordered by name</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>created (date)</b> - time the object was first created</li><li><b>id (string)</b> - unique object identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>metagenome (reference metagenome)</b> - reference to the related metagenome object</li><li><b>name (string)</b> - human readable identifier</li><li><b>project (reference project)</b> - reference to the project object</li><li><b>sample (reference sample)</b> - reference to the related sample object</li><li><b>sequencesets (list of reference sequencesets)</b> - a list of references to the related sequence sets</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>version (integer)</b> - version of the object</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns a standard subselection of metadata</li><li><b>full</b> - returns all connected metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/library/mgl52924?verbosity=full<li>retrieve all data for library mgl52924</ul><h3>Return Attributes</h3><ul><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p><li><b>metagenome</b> (reference metagenome)</li><p>reference to the related metagenome object</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>project</b> (reference project)</li><p>reference to the project object</p><li><b>sample</b> (reference sample)</li><p>reference to the related sample object</p><li><b>sequencesets</b> (list of reference sequencesets)</li><p>a list of references to the related sequence sets</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><hr><h1><a name='m5nr' style='padding-top: 50px;'>m5nr</a></h1><h2>Description</h2><p>M5NR provides data through a comprehensive non-redundant protein / rRNA database</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>ontology - request</h2><h3>Description</h3><p>Return functional hierarchy</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>filter</b> (string)</p><p>text of ontology group (filter_level) to filter by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom ontology level</li><li><b>level3</b> - ontology level</li><li><b>level2</b> - ontology level</li><li><b>level1</b> - top ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>min_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom ontology level</li><li><b>level3</b> - ontology level</li><li><b>level2</b> - ontology level</li><li><b>level1</b> - top ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type ontology only</li><li><b>NOG</b> - ontology database, type ontology only</li><li><b>COG</b> - ontology database, type ontology only</li><li><b>KO</b> - ontology database, type ontology only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/ontology?source=Subsystems&min_level=level3<li>retrieve subsystems hierarchy for the top 3 levels</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - ontology ID</li><li><b>level1 (string)</b> - ontology top level</li><li><b>level2 (string)</b> - ontology level 2</li><li><b>level3 (string)</b> - ontology level 3</li><li><b>level4 (string)</b> - ontology bottom level</li><li><b>source (string)</b> - source name</li></ul><br><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>taxonomy - request</h2><h3>Description</h3><p>Return organism hierarchy</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>filter</b> (string)</p><p>text of taxanomy group (filter_level) to filter by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>species</b> - taxonomy level</li><li><b>genus</b> - taxonomy level</li><li><b>family</b> - taxonomy level</li><li><b>order</b> -  taxonomy level</li><li><b>class</b> -  taxonomy level</li><li><b>phylum</b> - taxonomy level</li><li><b>domain</b> - top taxonomy level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>min_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>species</b> - taxonomy level</li><li><b>genus</b> - taxonomy level</li><li><b>family</b> - taxonomy level</li><li><b>order</b> -  taxonomy level</li><li><b>class</b> -  taxonomy level</li><li><b>phylum</b> - taxonomy level</li><li><b>domain</b> - top taxonomy level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/taxonomy?parent_name=Bacteroidetes&min_level=class<li>retrieve all class level taxa that belong to Bacteroidetes</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>class (string)</b> - organism class</li><li><b>domain (string)</b> - organism domain</li><li><b>family (string)</b> - organism family</li><li><b>genus (string)</b> - organism genus</li><li><b>ncbi_tax_id (int)</b> - organism ncbi id</li><li><b>order (string)</b> - organism order</li><li><b>organism (string)</b> - organism name</li><li><b>phylum (string)</b> - organism phylum</li><li><b>species (string)</b> - organism species</li></ul><br><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>sources - request</h2><h3>Description</h3><p>Return all sources in M5NR</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/sources<li>retrieve all data sources for M5NR</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>source object</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>accession - request</h2><h3>Description</h3><p>Return annotation or sequence of given source protein ID</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique identifier from source DB</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/accession/YP_003268079.1<li>retrieve M5NR data for accession ID 'YP_003268079.1'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>md5 - request</h2><h3>Description</h3><p>Return annotation(s) or sequence of given md5sum (M5NR ID)</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>sequence</b> (boolean)</p><p>if true return sequence output, else return annotation output, default is false</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique identifier in form of md5 checksum</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/md5/000821a2e2f63df1a3873e4b280002a8?source=InterPro<li>retrieve InterPro M5NR data for md5sum '000821a2e2f63df1a3873e4b280002a8'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>function - request</h2><h3>Description</h3><p>Return annotations for function names containing the given text</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>exact</b> (boolean)</p><p>if true return only those annotations that exactly match input text, default is false</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>text</b> (string)</p><p>text string of partial function name</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/function/sulfatase?source=GenBank<li>retrieve GenBank M5NR data for function names containing string 'sulfatase'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>organism - request</h2><h3>Description</h3><p>Return annotations for organism names containing the given text</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>exact</b> (boolean)</p><p>if true return only those annotations that exactly match input text, default is false</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>text</b> (string)</p><p>text string of partial organism name</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/organism/akkermansia?source=KEGG<li>retrieve KEGG M5NR data for organism names containing string 'akkermansia'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>sequence - request</h2><h3>Description</h3><p>Return annotation(s) for md5sum (M5NR ID) of given sequence</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>text</b> (string)</p><p>text string of protein sequence</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/m5nr/sequence/MAGENHQWQGSIL?source=TrEMBL<li>retrieve TrEMBL M5NR data for md5sum of sequence 'MAGENHQWQGSIL'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>accession - request</h2><h3>Description</h3><p>Return annotations of given source protein IDs</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>data</b> (list of string)</li><p>unique identifier from source DB</p><p>This parameter must be passed in the message body.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This parameter must be passed in the message body.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This parameter must be passed in the message body.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"order":"function","data":["YP_003268079.1","COG1764"]}' "http://kbase.us/services/communities/1/m5nr/accession"<li>retrieve M5NR data for accession IDs 'YP_003268079.1' and 'COG1764' ordered by function</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>md5 - request</h2><h3>Description</h3><p>Return annotations of given md5sums (M5NR ID)</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>data</b> (list of string)</li><p>unique identifier in form of md5 checksum</p><p>This parameter must be passed in the message body.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This parameter must be passed in the message body.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This parameter must be passed in the message body.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This parameter must be passed in the message body.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"source":"InterPro","data":["000821a2e2f63df1a3873e4b280002a8","15bf1950bd9867099e72ea6516e3d602"]}' "http://kbase.us/services/communities/1/m5nr/md5"<li>retrieve InterPro M5NR data for md5s '000821a2e2f63df1a3873e4b280002a8' and '15bf1950bd9867099e72ea6516e3d602'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>function - request</h2><h3>Description</h3><p>Return annotations for function names containing the given texts</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>data</b> (list of string)</li><p>text string of partial function name</p><p>This parameter must be passed in the message body.</p><li><b>exact</b> (boolean)</p><p>if true return only those annotations that exactly match input text, default is false</p><p>This parameter must be passed in the message body.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This parameter must be passed in the message body.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This parameter must be passed in the message body.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This parameter must be passed in the message body.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"source":"GenBank","limit":50,"data":["sulfatase","phosphatase"]}' "http://kbase.us/services/communities/1/m5nr/function"<li>retrieve top 50 GenBank M5NR data for function names containing string 'sulfatase' or 'phosphatase'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>organism - request</h2><h3>Description</h3><p>Return annotations for organism names containing the given texts</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>data</b> (list of string)</li><p>text string of partial organism name</p><p>This parameter must be passed in the message body.</p><li><b>exact</b> (boolean)</p><p>if true return only those annotations that exactly match input text, default is false</p><p>This parameter must be passed in the message body.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This parameter must be passed in the message body.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This parameter must be passed in the message body.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This parameter must be passed in the message body.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"source":"KEGG","order":"accession","data":["akkermansia","yersinia"]}' "http://kbase.us/services/communities/1/m5nr/organism"<li>retrieve KEGG M5NR data (ordered by accession ID) for organism names containing string 'akkermansia' or 'yersinia'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>sequence - request</h2><h3>Description</h3><p>Return annotations for md5s (M5NR ID) of given sequences</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>data</b> (list of string)</li><p>text string of protein sequence</p><p>This parameter must be passed in the message body.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This parameter must be passed in the message body.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This parameter must be passed in the message body.</p><li><b>order</b> (string)</p><p>name of the attribute the returned data is ordered by</p><p>This parameter must be passed in the message body.</p><li><b>source</b> (string)</p><p>source name to restrict search by</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -d '{"source":"KEGG","order":"source","data":["MAGENHQWQGSIL","MAGENHQWQGSIL"]}' "http://kbase.us/services/communities/1/m5nr/sequence"<li>retrieve M5NR data ordered by source for sequences 'MAGENHQWQGSIL' and 'MAGENHQWQGSIL'</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>accession (string)</b> - unique identifier given by source</li><li><b>function (string)</b> - function annotation</li><li><b>md5 (string)</b> - md5 checksum - M5NR ID</li><li><b>ncbi_tax_id (int)</b> - organism ncbi tax_id</li><li><b>organism (string)</b> - organism annotation</li><li><b>source (string)</b> - source name</li><li><b>type (string)</b> - source type</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><hr><h1><a name='matrix' style='padding-top: 50px;'>matrix</a></h1><h2>Description</h2><p>A profile in biom format that contains abundance counts</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>organism - request</h2><h3>Description</h3><p>Returns a BIOM object.</p><p>This is a synchronous or asynchronous GET request.</p><h3>Parameters</h3><ul><li><b>asynchronous</b> (boolean)</p><p>if true, return process id to query status resource for results.  default is false.</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</p><p>negative exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</p><p>filter the return results to only include abundances based on genes with this function</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom function ontology level</li><li><b>level3</b> - function ontology level</li><li><b>level2</b> - function ontology level</li><li><b>level1</b> - top function ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type ontology only</li><li><b>NOG</b> - ontology database, type ontology only</li><li><b>COG</b> - ontology database, type ontology only</li><li><b>KO</b> - ontology database, type ontology only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>grep</b> (string)</p><p>filter the return results to only include annotations that contain this text</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>strain</b> - bottom organism taxanomic level</li><li><b>species</b> - organism type level</li><li><b>genus</b> - organism taxanomic level</li><li><b>family</b> - organism taxanomic level</li><li><b>order</b> - organism taxanomic level</li><li><b>class</b> - organism taxanomic level</li><li><b>phylum</b> - organism taxanomic level</li><li><b>domain</b> - top organism taxanomic level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</p><p>if false, return metagenome metadata set in 'columns' object.  default is false.</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hit_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>all</b> - returns results based on all organisms that map to top hit per read-feature</li><li><b>single</b> - returns results based on a single organism for top hit per read-feature</li><li><b>lca</b> - returns results based on the Least Common Ancestor for all organisms (M5NR+M5RNA only) that map to hits from a read-feature</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</p><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</p><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RefSeq</b> - protein database, type organism, function, feature</li><li><b>GenBank</b> - protein database, type organism, function, feature</li><li><b>IMG</b> - protein database, type organism, function, feature</li><li><b>SEED</b> - protein database, type organism, function, feature</li><li><b>TrEMBL</b> - protein database, type organism, function, feature</li><li><b>SwissProt</b> - protein database, type organism, function, feature</li><li><b>PATRIC</b> - protein database, type organism, function, feature</li><li><b>KEGG</b> - protein database, type organism, function, feature</li><li><b>RDP</b> - RNA database, type organism, function, feature</li><li><b>Greengenes</b> - RNA database, type organism, function, feature</li><li><b>LSU</b> - RNA database, type organism, function, feature</li><li><b>SSU</b> - RNA database, type organism, function, feature</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>taxid</b> (boolean)</p><p>if true, return annotation ID as NCBI tax id. Only for group_levels with a tax_id</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/matrix/organism?id=mgm4447943.3&id=mgm4447192.3&id=mgm4447102.3&id=mgm4447103.3&group_level=family&source=RefSeq&result_type=abundance&evalue=15<li>retrieve abundance matrix of RefSeq organism annotations at family taxa for listed metagenomes at evaule < e-15</ul><h3>Return Attributes</h3><ul><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p></p></ul><h2>function - request</h2><h3>Description</h3><p>Returns a BIOM object.</p><p>This is a synchronous or asynchronous GET request.</p><h3>Parameters</h3><ul><li><b>asynchronous</b> (boolean)</p><p>if true, return process id to query status resource for results.  default is false.</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</p><p>negative exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</p><p>filter the return results to only include abundances based on genes with this organism</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>strain</b> - bottom organism taxanomic level</li><li><b>species</b> - organism type level</li><li><b>genus</b> - organism taxanomic level</li><li><b>family</b> - organism taxanomic level</li><li><b>order</b> - organism taxanomic level</li><li><b>class</b> - organism taxanomic level</li><li><b>phylum</b> - organism taxanomic level</li><li><b>domain</b> - top organism taxanomic level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RefSeq</b> - protein database, type organism, function, feature</li><li><b>GenBank</b> - protein database, type organism, function, feature</li><li><b>IMG</b> - protein database, type organism, function, feature</li><li><b>SEED</b> - protein database, type organism, function, feature</li><li><b>TrEMBL</b> - protein database, type organism, function, feature</li><li><b>SwissProt</b> - protein database, type organism, function, feature</li><li><b>PATRIC</b> - protein database, type organism, function, feature</li><li><b>KEGG</b> - protein database, type organism, function, feature</li><li><b>RDP</b> - RNA database, type organism, function, feature</li><li><b>Greengenes</b> - RNA database, type organism, function, feature</li><li><b>LSU</b> - RNA database, type organism, function, feature</li><li><b>SSU</b> - RNA database, type organism, function, feature</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>grep</b> (string)</p><p>filter the return results to only include annotations that contain this text</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom function ontology level</li><li><b>level3</b> - function ontology level</li><li><b>level2</b> - function ontology level</li><li><b>level1</b> - top function ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</p><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</p><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</p><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type ontology only</li><li><b>NOG</b> - ontology database, type ontology only</li><li><b>COG</b> - ontology database, type ontology only</li><li><b>KO</b> - ontology database, type ontology only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/matrix/function?id=mgm4447943.3&id=mgm4447192.3&id=mgm4447102.3&id=mgm4447103.3&group_level=level3&source=Subsystems&result_type=abundance&identity=80&filter_level=phylum&filter=Firmicutes<li>retrieve abundance matrix of Subsystem annotations at level3 for listed metagenomes at % identity > 80, filtered to return annotations only in phylum Firmicutes</ul><h3>Return Attributes</h3><ul><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p></p></ul><h2>feature - request</h2><h3>Description</h3><p>Returns a BIOM object.</p><p>This is a synchronous or asynchronous GET request.</p><h3>Parameters</h3><ul><li><b>asynchronous</b> (boolean)</p><p>if true, return process id to query status resource for results.  default is false.</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</p><p>negative exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>grep</b> (string)</p><p>filter the return results to only include annotations that contain this text</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</p><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</p><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</p><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RefSeq</b> - protein database, type organism, function, feature</li><li><b>GenBank</b> - protein database, type organism, function, feature</li><li><b>IMG</b> - protein database, type organism, function, feature</li><li><b>SEED</b> - protein database, type organism, function, feature</li><li><b>TrEMBL</b> - protein database, type organism, function, feature</li><li><b>SwissProt</b> - protein database, type organism, function, feature</li><li><b>PATRIC</b> - protein database, type organism, function, feature</li><li><b>KEGG</b> - protein database, type organism, function, feature</li><li><b>RDP</b> - RNA database, type organism, function, feature</li><li><b>Greengenes</b> - RNA database, type organism, function, feature</li><li><b>LSU</b> - RNA database, type organism, function, feature</li><li><b>SSU</b> - RNA database, type organism, function, feature</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/matrix/feature?id=mgm4447943.3&id=mgm4447192.3&id=mgm4447102.3&id=mgm4447103.3&source=KEGG&result_type=evalue&length=25<li>retrieve e-value matrix of KEGG protein annotations for listed metagenomes at alignment length > 25</ul><h3>Return Attributes</h3><ul><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p></p></ul><hr><h1><a name='metadata' style='padding-top: 50px;'>metadata</a></h1><h2>Description</h2><p>Metagenomic metadata is data providing information about one or more aspects of a set sequences from a sample of some environment</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>template - request</h2><h3>Description</h3><p>Returns static template for metadata object relationships and types</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/metadata/template<li>metadata template</ul><h3>Return Attributes</h3><ul><li><b>ep</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - enviromental package type</li><li><b>value (hash)</b> - hash of metadata objects by label</li></ul><br><li><b>library</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - library type</li><li><b>value (hash)</b> - hash of metadata objects by label</li></ul><br><li><b>project</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - project type</li><li><b>value (hash)</b> - hash of metadata objects by label</li></ul><br><li><b>sample</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - sample type</li><li><b>value (hash)</b> - hash of metadata objects by label</li></ul><br></p></ul><h2>cv - request</h2><h3>Description</h3><p>Returns static controlled vocabularies used in metadata</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/metadata/cv<li>metadata controlled vobularies</ul><h3>Return Attributes</h3><ul><li><b>ont_id</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - metadata label</li><li><b>value (int)</b> - bioportal term ID</li></ul><br><li><b>ontology</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - metadata label</li><li><b>value (int)</b> - bioportal ontology ID</li></ul><br><li><b>select</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - metadata label</li><li><b>value (list of strings)</b> - CV term</li></ul><br></p></ul><h2>export - request</h2><h3>Description</h3><p>Returns full nested metadata for a project in same format as template</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>id</b> (string)</p><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/metadata/export/mgp128<li>all metadata for project mgp128</ul><h3>Return Attributes</h3><ul><li><b>data</b> (hash)</li><p>This attribute has a hash structure:</p><ul style='list-style: none;'><li><b>key (string)</b> - metadata label</li><li><b>value (object)</b> - project metadata objects</li></ul><br><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>sampleNum</b> (int)</li><p>number of samples in project</p><li><b>samples</b> (list of objects)</li><p>sample object containing sample metadata, sample libraries, sample envPackage</p></p></ul><h2>validate - request</h2><h3>Description</h3><p>Validate given metadata spreadsheet</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>upload</b> (file)</p><p>xlsx or xls format spreadsheet with metadata</p><p>This parameter must be passed in the message body.</p></ul><h3>Example</h3><ul><li>curl -X POST -F "upload=@metadata.xlsx" "http://kbase.us/services/communities/1/metadata/validate"<li>validate file 'metadata.xlsx' against MG-RAST metadata template</ul><h3>Return Attributes</h3><ul><li><b>is_valid</b> (boolean)</li><p>the metadata sheet is valid</p><li><b>message</b> (string)</li><p>if not valid, reason why</p><li><b>metadata</b> (object)</li><p>valid metadata object for project and its samples and libraries</p></p></ul><h2>validate - request</h2><h3>Description</h3><p>Validate given metadata value</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>category</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>project</b> - label belongs to project metadata</li><li><b>sample</b> - label belongs to sample metadata</li><li><b>library</b> - label belongs to library metadata</li><li><b>env_package</b> - label belongs to env_package metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>group</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>mixs</b> - label is part of MIxS (minimal) metadata</li><li><b>mims</b> - label is part of MIMS (metagenome) metadata</li><li><b>migs</b> - label is part of MIGS (genome) metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>label</b> (string)</p><p>metadata label</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>value</b> (string)</p><p>metadata value</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/metadata/validate?category=sample&label=material&value=soil<li>check if 'soil' is a vaild term for sample material</ul><h3>Return Attributes</h3><ul><li><b>is_valid</b> (boolean)</li><p>the inputed value is valid for the given category and label</p><li><b>message</b> (string)</li><p>if not valid, reason why</p></p></ul><hr><h1><a name='metagenome' style='padding-top: 50px;'>metagenome</a></h1><h2>Description</h2><p>A metagenome is an analyzed set sequences from a sample of some environment</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>direction</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>asc</b> - sort by ascending order</li><li><b>desc</b> - sort by descending order</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>function</b> (string)</p><p>search parameter: query string for function</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>match</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>all</b> - return metagenomes that match all search parameters</li><li><b>any</b> - return metagenomes that match any search parameters</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>md5</b> (string)</p><p>search parameter: md5 checksum of feature sequence</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>metadata</b> (string)</p><p>search parameter: query string for any metadata field</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (string)</p><p>metagenome object field to sort by (default is id)</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>organism</b> (string)</p><p>search parameter: query string for organism</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>status</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>both</b> - returns all data (public and private) user has access to view</li><li><b>public</b> - returns all public data</li><li><b>private</b> - returns private data user has access to view</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>mixs</b> - returns all GSC MIxS metadata</li><li><b>metadata</b> - returns minimal with metadata</li><li><b>stats</b> - returns minimal with statistics</li><li><b>full</b> - returns all metadata and statistics</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/metagenome?limit=20&order=name<li>retrieve the first 20 metagenomes ordered by name</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>PI_firstname (string)</b> - principal investigator's first name</li><li><b>PI_lastname (string)</b> - principal investigator's last name</li><li><b>biome (string)</b> - environmental biome, EnvO term</li><li><b>collection_date (string)</b> - date sample collected</li><li><b>country (string)</b> - country where sample taken</li><li><b>created (date)</b> - time the metagenome was first created</li><li><b>env_package_type (string)</b> - enviromental package of sample, GSC term</li><li><b>feature (string)</b> - environmental feature, EnvO term</li><li><b>id (string)</b> - unique metagenome identifier</li><li><b>latitude (string)</b> - latitude where sample taken</li><li><b>location (string)</b> - location where sample taken</li><li><b>longitude (string)</b> - longitude where sample taken</li><li><b>material (string)</b> - environmental material, EnvO term</li><li><b>name (string)</b> - name of metagenome</li><li><b>project_id (string)</b> - id of project containing metagenome</li><li><b>project_name (string)</b> - name of project containing metagenome</li><li><b>seq_method (string)</b> - sequencing method</li><li><b>sequence_type (string)</b> - sequencing type</li><li><b>status (cv of ARRAY(0x7fd589a9e098)s of privates)</b> - metagenome is private</li><li><b>url (uri)</b> - resource location of this object instance</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>metadata</b> - returns minimal with metadata</li><li><b>stats</b> - returns minimal with statistics</li><li><b>full</b> - returns all metadata and statistics</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/metagenome/mgm4447943.3?verbosity=metadata<li>retrieve all metadata for metagenome mgm4447943.3</ul><h3>Return Attributes</h3><ul><li><b>created</b> (date)</li><p>time the metagenome was first created</p><li><b>id</b> (string)</li><p>unique metagenome identifier</p><li><b>library</b> (reference library)</li><p>reference to the related library object</p><li><b>metadata</b> (hash)</li><p>key value pairs describing all metadata</p><li><b>mixs</b> (hash)</li><p>key value pairs describing MIxS metadata</p><li><b>name</b> (string)</li><p>name of metagenome</p><li><b>project</b> (reference project)</li><p>reference to the project object</p><li><b>sample</b> (reference sample)</li><p>reference to the related sample object</p><li><b>sequence_type</b> (string)</li><p>sequencing type</p><li><b>statistics</b> (hash)</li><p>key value pairs describing statistics</p><li><b>status</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>public</b> - metagenome is public</li><li><b>private</b> - metagenome is private</li></ul><br><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the metagenome</p></p></ul><hr><h1><a name='project' style='padding-top: 50px;'>project</a></h1><h2>Description</h2><p>A project is a composition of samples, libraries and metagenomes being analyzed in a global context.</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns all metadata</li><li><b>full</b> - returns all metadata and references</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/project?limit=20&order=name<li>retrieve the first 20 projects ordered by name</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>created (date)</b> - time the object was first created</li><li><b>description (string)</b> - a short, comprehensive description of the project</li><li><b>funding_source (string)</b> - the official name of the source of funding of this project</li><li><b>id (string)</b> - unique object identifier</li><li><b>libraries (list of reference librarys)</b> - a list of references to the related library objects</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>metagenomes (list of reference metagenomes)</b> - a list of references to the related metagenome objects</li><li><b>name (string)</b> - human readable identifier</li><li><b>pi (string)</b> - the first and last name of the principal investigator of the project</li><li><b>samples (list of reference samples)</b> - a list of references to the related sample objects</li><li><b>status (cv of ARRAY(0x7fd589a90530)s of privates)</b> - object is private</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>version (integer)</b> - version of the object</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns all metadata</li><li><b>full</b> - returns all metadata and references</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/project/mgp128?verbosity=full<li>retrieve all data for project mgp128</ul><h3>Return Attributes</h3><ul><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>description</b> (string)</li><p>a short, comprehensive description of the project</p><li><b>funding_source</b> (string)</li><p>the official name of the source of funding of this project</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>libraries</b> (list of reference librarys)</li><p>a list of references to the related library objects</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p><li><b>metagenomes</b> (list of reference metagenomes)</li><p>a list of references to the related metagenome objects</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>pi</b> (string)</li><p>the first and last name of the principal investigator of the project</p><li><b>samples</b> (list of reference samples)</li><p>a list of references to the related sample objects</p><li><b>status</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>public</b> - object is public</li><li><b>private</b> - object is private</li></ul><br><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><hr><h1><a name='sample' style='padding-top: 50px;'>sample</a></h1><h2>Description</h2><p>A metagenomic sample from some environment.</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>limit</b> (integer)</p><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</p><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/sample?limit=20&order=name<li>retrieve the first 20 samples ordered by name</ul><h3>Return Attributes</h3><ul><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>created (date)</b> - time the object was first created</li><li><b>env_package (object of HASH(0x7fd589ab5af0)s)</b> - environmental package object</li><li><b>id (string)</b> - unique object identifier</li><li><b>libraries (list of reference librarys)</b> - a list of references to the related library objects</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>metagenomes (list of reference metagenomes)</b> - a list of references to the related metagenome objects</li><li><b>name (string)</b> - human readable identifier</li><li><b>project (reference project)</b> - reference to the project of this sample</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>version (integer)</b> - version of the object</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns all metadata</li><li><b>full</b> - returns all metadata and references</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</p><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Example</h3><ul><li>http://kbase.us/services/communities/1/sample/mgs25823?verbosity=full<li>retrieve all data for sample mgs25823</ul><h3>Return Attributes</h3><ul><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>env_package</b> (object)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>created (date)</b> - creation date</li><li><b>id (string)</b> - unique package identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>name (string)</b> - name of the package</li><li><b>type (string)</b> - package type</li></ul><br><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>libraries</b> (list of reference librarys)</li><p>a list of references to the related library objects</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p><li><b>metagenomes</b> (list of reference metagenomes)</li><p>a list of references to the related metagenome objects</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>project</b> (reference project)</li><p>reference to the project of this sample</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>version</b> (integer)</li><p>version of the object</p></p></ul><hr>
</ul>
    </div>
  </body>

</html>
